### PR TITLE
DatePicker: Fix calendar not showing correct selected range when changing time zones

### LINKED ISSUE #69939: Timepicker: Datepicker for absolute time range adding extra day
### 
# What went wrong?

The color overlay in the datepicker is one day longer than what is selected by clicking on the dates. The "To" date textbox correctly reflects the date that was clicked on. This happens for any interval length (in days).
When klicking a single day repeatedly the overlay switches between one and two days.
I can reproduce this in Edge and Firefox.

grafik

### 
# What Grafana version are you using?

9.5.2

### Grafana Platform?

A package manager (APT, YUM, BREW, etc.)

### PR DESCRIPTION
**Pull request by ashharrison90**
Fixes https://github.com/grafana/grafana/issues/69939

Fixes the Date Picker calendar not showing the correct selected dates when the user's timezone Grafana preference is different to their browser/OS.
    - [react-calendar doesn't support time zones](https://github.com/wojtekmaj/react-calendar/issues/511#issuecomment-835333976), so we must fiddle with the date to make them 'appear' correct to react-calendar. e.g. If 5 PM in New York is being selected we change the date to 5 PM in the user's browser's local time zone. 

Added E2E tests for the above, which required adding a new flow to set user preferences, and reset them when the test is finished.

### CODE REVIEW

File: packages/grafana-e2e/src/support/scenarioContext.ts [hunk 1]
```diff
@@ -45,8 +46,7 @@ export const getScenarioContext = (): any =>
     )
     .invoke({ log: false }, 'getScenarioContext');
 
-// @todo this actually returns type `Cypress.Chainable`
-export const setScenarioContext = (newContext: Partial<ScenarioContext>): any =>
+export const setScenarioContext = (newContext: Partial<ScenarioContext>): Cypress.Chainable<ScenarioContext> =>
```
Comment (ashharrison90) [inline packages/grafana-e2e/src/support/scenarioContext.ts:49]: think you need to run betterer to update the results file

**Review by ashharrison90 — COMMENTED** [review]
can't approve a PR i'm the author of... but looks great!

minor comment otherwise CI will fail

**Review by JoaoSilvaGrafana — APPROVED** [review]
nice! great job on the tests as well!!

**Review by mckn — COMMENTED** [review]
I have spent some time with the time picker + time zones previously but I have not followed the latest updates. So what I'm suggesting might not be applicable but shouldn't it be enough to always convert the date to UTC prior to passing it to the calendar. Then always convert from UTC when reading the values from the calendar.

As you said, the calendar doesn't support time zones so if we always work with UTC it shouldn't be needed.

File: packages/grafana-data/src/datetime/timezones.ts [hunk 2]
```diff
@@ -22,6 +22,10 @@ export const timeZoneFormatUserFriendly = (timeZone: TimeZone | undefined) => {
   }
 };
 
+export const getZone = (timeZone: string) => {
+  return moment.tz.zone(timeZone);
```
Comment (mckn) [inline packages/grafana-data/src/datetime/timezones.ts:26]: Does this one support the "grafana specific" time zones? `browser` and I think there should be another one...hmm..don't remember 🙈

**Comment by ashharrison90** [comment]
hey @mckn 👋 

converting to UTC will show a different range unless your browser local time is also UTC. as mentioned `react-calendar` doesn't support timezones, but always shows the times **in your local browser timezone**. let's take an example:

- you're in sweden (UTC+2)
- you set your grafana timezone to tokyo (UTC+9)
- you set the time range to be 2023-08-16 00:00:00 to 2023-08-17 23:59:59

## what happens now:
  - `react-calendar` gets passed the times and converts them to the browser timezone
  - in the browser timezone, those times become 2023-08-15 17:00:00 to 2023-08-17 16:59:59
  - `react-calendar` shows the 15th, 16th and 17th highlighted.
  - confusing, because it doesn't match what's shown in the input boxes

## what would happen if we just converted to UTC
  - in UTC, those times become 2023-08-15 15:00:00 to 2023-08-17 14:59:59
  - `react-calendar` gets passed the times and converts them to the browser timezone
  - in the browser timezone, those times become 2023-08-15 17:00:00 to 2023-08-17 16:59:59
  - `react-calendar` shows the 15th, 16th and 17th highlighted.
  - confusing, because it doesn't match what's shown in the input boxes

## what happens after this PR:
  - we calculate the offset between grafana's timezone and the browser timezone (7 in this case)
  - we adjust the times themselves in grafana's timezone before passing them to `react-calendar`
  - the times become 2023-08-16 07:00:00 to 2023-08-18 06:59:59 (still in tokyo timezone)
  - **this is technically the wrong time range**. but it doesn't matter - we're only using it so that `react-calendar` displays correctly
  - `react-calendar` gets passed the times and converts them to the browser timezone
  - in the browser timezone, those times become 2023-08-16 00:00:00 to 2023-08-17 23:59:59
  - `react-calendar` shows the 16th and 17th as highlighted which matches the input boxes 🥳 

re: `getZone` - `browser` is handled correctly. `getZone('browser')` will return `null`, which will mean no adjustment takes place which is correct. it would be good to know what the other internal grafana zone can be as that needs checking if it works.

**Comment by mckn** [comment]
> hey @mckn 👋
> 
> converting to UTC will show a different range unless your browser local time is also UTC. as mentioned `react-calendar` doesn't support timezones, but always shows the times **in your local browser timezone**. let's take an example:
> 
> * you're in sweden (UTC+2)
> * you set your grafana timezone to tokyo (UTC+9)
> * you set the time range to be 2023-08-16 00:00:00 to 2023-08-17 23:59:59
> 
> ## what happens now:
> * `react-calendar` gets passed the times and converts them to the browser timezone
> * in the browser timezone, those times become 2023-08-15 17:00:00 to 2023-08-17 16:59:59
> * `react-calendar` shows the 15th, 16th and 17th highlighted.
> * confusing, because it doesn't match what's shown in the input boxes
> 
> ## what would happen if we just converted to UTC
> * in UTC, those times become 2023-08-15 15:00:00 to 2023-08-17 14:59:59
> * `react-calendar` gets passed the times and converts them to the browser timezone
> * in the browser timezone, those times become 2023-08-15 17:00:00 to 2023-08-17 16:59:59
> * `react-calendar` shows the 15th, 16th and 17th highlighted.
> * confusing, because it doesn't match what's shown in the input boxes
> 
> ## what happens after this PR:
> * we calculate the offset between grafana's timezone and the browser timezone (7 in this case)
> * we adjust the times themselves in grafana's timezone before passing them to `react-calendar`
> * the times become 2023-08-16 07:00:00 to 2023-08-18 06:59:59 (still in tokyo timezone)
> * **this is technically the wrong time range**. but it doesn't matter - we're only using it so that `react-calendar` displays correctly
> * `react-calendar` gets passed the times and converts them to the browser timezone
> * in the browser timezone, those times become 2023-08-16 00:00:00 to 2023-08-17 23:59:59
> * `react-calendar` shows the 16th and 17th as highlighted which matches the input boxes 🥳
> 
> re: `getZone` - `browser` is handled correctly. `getZone('browser')` will return `null`, which will mean no adjustment takes place which is correct. it would be good to know what the other internal grafana zone can be as that needs checking if it works.

Ah, got it! Thanks for the amazing explanation. I guess you could use the same logic but always converting to the browser time zone. But I guess that is what you do?

**Pull request merged**